File: LeetCode/918.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int> &nums) {
        int n = nums.size(), INF = std::numeric_limits<int>::max();

        std::vector<int> preMax(n + 1);
        preMax[0] = -INF;
        int max_1 = -INF, max_curr = 0, pre = 0;
        for (int i = 0; i < n; i++) {
            pre += nums[i];
            preMax[i + 1] = std::max(preMax[i], pre);
            max_curr = std::max(max_curr + nums[i], nums[i]);
            max_1 = std::max(max_1, max_curr);
        }

        int max_2 = -INF, suf = 0;
        for (int i = n - 1; i > 0; i--) {
            suf += nums[i];
            max_2 = std::max(max_2, suf + preMax[i]);
        }

        return std::max(max_1, max_2);
    }
};

File: LeetCode/918.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        dp_max = -0x3F3F3F3F
        curr_max = 0
        dp_min = 0x3F3F3F3F
        curr_min = 0
        s = 0
        for elem in nums:
            curr_max = max(curr_max, 0) + elem
            curr_min = min(curr_min, 0) + elem
            dp_max = max(dp_max, curr_max)
            dp_min = min(dp_min, curr_min)
            s += elem

        return dp_max if dp_min == s else max(dp_max, s - dp_min)
